### PR TITLE
feat(tiering): Faster small bins serialization

### PR DESCRIPTION
**Pull request by dranikpg**

### CODE REVIEW

File: src/server/snapshot.cc [hunk 1]
```diff
@@ -37,8 +38,80 @@ size_t SliceSnapshot::DbRecord::size() const {
...
+      }
+    } else {
+      pending_.push_back(Read(std::move(base), pv));
+    }
+  }
+
+  void FlushDelayed(bool force) {
+    // Flush pages with most records accumulated first, or all, if forced.
+    // It's enough just to issue reads, because they are collapsed by the tiered storage internally
+    while ((force && !delayed_.empty()) || delayed_.size() > kMaxPageAccum) {
```
Comment (dranikpg) [inline src/server/snapshot.cc:76]: We need to choose some kind of limit here... by the number of pages? by the total sum of enties? 🤷🏻‍♂️ We have to take into account that once we flush all of those, we can get a memory spike

File: src/server/snapshot.cc [hunk 2]
```diff
@@ -377,6 +435,9 @@ void SliceSnapshot::OnDbChange(DbIndex db_index, const DbSlice::ChangeReq& req)
       stats_.side_saved += SerializeBucket(db_index, it);
     });
   }
+
+  // Flush tiered delayed entries to avoid reordering with journal
+  tiered_serializer_->FlushDelayed(true);
 }
```
Comment (dranikpg) [inline src/server/snapshot.cc:441]: Ideally we pass the value with ChangeReq& and find out whether it was tiered, only if it's tiered and part of a relevant page, we should flush it

File: src/server/tiered_storage.cc [hunk 3]
```diff
@@ -158,7 +158,14 @@ class TieredStorage::ShardOpManager : public tiering::OpManager {
...
   }
 
+  void ClearDelayed() {
+    for (auto segment : delayed_deletes_)
+      OpManager::DeleteOffloaded(segment);
+    delayed_deletes_.clear();
+  }
+
   int64_t memory_margin_ = 0;
+  std::vector<tiering::DiskSegment> delayed_deletes_;
```
Comment (dranikpg) [inline src/server/tiered_storage.cc:168]: todo: better naming

**Comment by dranikpg** [comment]
@romange Please look only at the idea, I left some open questions

File: src/server/snapshot.cc [hunk 4]
```diff
@@ -37,8 +38,80 @@ size_t SliceSnapshot::DbRecord::size() const {
...
+  absl::flat_hash_map<size_t /* page */, std::vector<DelayedValue>> delayed_;
+  // Largest entries in delayed map
+  absl::btree_set<std::pair<size_t /* size */, size_t /* page */>, std::greater<>> delayed_sizes_;
+};
+
 SliceSnapshot::SliceSnapshot(DbSlice* slice, RecordChannel* dest, CompressionMode compression_mode)
-    : db_slice_(slice), dest_(dest), compression_mode_(compression_mode) {
+    : db_slice_(slice),
+      dest_(dest),
+      tiered_serializer_(new TieredSerializer{}),
```
Comment (romange) [inline src/server/snapshot.cc:113]: do you want to initialize it unconditionally?

File: src/server/snapshot.cc [hunk 5]
```diff
@@ -37,8 +38,80 @@ size_t SliceSnapshot::DbRecord::size() const {
...
+  void FlushDelayed(bool force) {
+    // Flush pages with most records accumulated first, or all, if forced.
+    // It's enough just to issue reads, because they are collapsed by the tiered storage internally
+    while ((force && !delayed_.empty()) || delayed_.size() > kMaxPageAccum) {
+      DCHECK(!delayed_sizes_.empty());
+      auto [size, page] = delayed_sizes_.extract(delayed_sizes_.begin()).value();
+      auto entries = delayed_.extract(page);
+      for (auto& [base, value] : entries.mapped()) {
+        DCHECK(value.IsExternal());
+        pending_.push_back(Read(std::move(base), value));
```
Comment (romange) [inline src/server/snapshot.cc:82]: what guarantess that by the time you flush delayed, their segment are still correct?
maybe the pages were freed and repurposed for other values?

File: src/server/snapshot.cc [hunk 6]
```diff
@@ -37,8 +38,80 @@ size_t SliceSnapshot::DbRecord::size() const {
...
+      pv_copy.ImportExternal(pv);
+      size_t page = offset / tiering::kPageSize;
+      auto& entries = delayed_[page];
+      {
+        delayed_sizes_.erase({entries.size(), page});
+        entries.emplace_back(std::move(base), std::move(pv_copy));
+        delayed_sizes_.insert({entries.size(), page});
+      }
+    } else {
+      pending_.push_back(Read(std::move(base), pv));
```
Comment (romange) [inline src/server/snapshot.cc:69]: nit: pending_reads_

**Review by romange — COMMENTED** [review]
Nice that you suggest to do it without changing the RDB format.
I am not sure about the effectiveness of flush though. the probability of keys from the same page being in the same serialization batch are low.

File: src/server/snapshot.cc (hunk 5 above)
Comment (dranikpg) [reply src/server/snapshot.cc:82]: Inside tiered storage, I don't delete small bin pages while we're serializing

File: src/server/snapshot.cc (hunk 4 above)
Comment (dranikpg) [reply src/server/snapshot.cc:113]: 🤷🏻‍♂️ I don't think it's expensive

**Comment by romange** [comment]
I thought about this more. Master is the more fragile piece compared to slave or an instance loading the snapshot.
That means we want to move the complexity to Load phase and I think it is possible to do if we just send the segment support at the end. This suggestion tracks pending items on the master side, i.e. maintains per object  granularity.  I think it's worth extending RDB support to maintain segment granularity (i.e. send 4K pages with their offsets.).

**Comment by dranikpg** [comment]
Closed in favour of #3396

**Pull request closed without merging**